### PR TITLE
Updates wallet-js functionality, fix some bugs

### PR DESCRIPTION
**Pull request by Mr-Leshiy**
PRs:
- https://github.com/input-output-hk/catalyst-core/pull/245
- https://github.com/input-output-hk/catalyst-core/pull/249

### CODE REVIEW

File: src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts [hunk 1]
```diff
@@ -58,27 +62,44 @@ export class Proposal {
...
+  spendingCounter?: number;
+  spendingCounterLane?: number;
 
   /**
    * Vote constructor
    *
    * @param {Proposal} proposal
-   * @param {number} choice choosen vote plan option
+   * @param {number} choice choosen vote plan option.
+   * @param {VotingPurpose} purpose The voting purpose being voted on (Currently not used actually, can pass anything).
```
Comment (minikin) [inline src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts:75]: Additionally, we need to be consistent with comments TS/JS. We jump between starting comments from capital letters to regular: `choosen vote plan option.` - `The voting purpose being ...`. I'd propose always starting comments with a capital letter.

**Review by minikin — CHANGES_REQUESTED** [review]

File: src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts [hunk 2]
```diff
@@ -58,27 +62,44 @@ export class Proposal {
...
+  expiration?: BlockDate;
+  spendingCounter?: number;
+  spendingCounterLane?: number;
 
   /**
    * Vote constructor
    *
    * @param {Proposal} proposal
-   * @param {number} choice choosen vote plan option
+   * @param {number} choice choosen vote plan option.
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts:74]: I have followed a CIP-62 description https://github.com/ehanoc/CIPs/blob/cip-governance-wallet-connector/CIP-0062/README.md#vote

File: src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs [hunk 3]
```diff
@@ -513,14 +513,19 @@ impl FragmentReplayer {
             .get_mut(&address)
             .ok_or_else(|| ReplayError::NonVotingAccount(address.to_string()))?;
 
+        let secret_key = wallet.secret_key();
         // unwrap checked in the validation step
         let builder_help = wallet
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs:518]: Actually `tx.total_input()` has a `Result<Value, ValueError>` return value.
It is a good question. I am not deeply understood how `replay_votecast` function works, but as I see we are passing an already builded vote cast transaction, so seems that we are assuming that this transaction is correct at this place.

File: src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts [hunk 4]
```diff
@@ -58,27 +62,44 @@ export class Proposal {
...
+    choice: number,
+    purpose: VotingPurpose,
+    expiration?: BlockDate,
+    spendingCounter?: number,
+    spendingCounterLane?: number
+  ): Vote;
 }
 
 /**
  * Signes provided votes and returns a completly generated transaction list
```
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts]: fixed

File: src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (minikin) [reply src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts]: We don't need to follow that CIP "style guide" when someone failed to follow best practices.
@stevenj what do you think?

File: src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts (hunk 1 above)
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts:75]: fixed

File: src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (minikin) [reply src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs:518]: But we still use `unwrap()` twice to allocate `builder_help` and we definitely don't want to `panic`.
Should we consider using `[unwrap_or], [unwrap_or_else], or [unwrap_or_default].` instead, at least 🧐?

[the remark this answers is not in the document]
Comment (minikin) [reply src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs:518]: ideally, we need to handle the error state properly.

File: src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts]: fixed

File: src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/chain-wallet-libs/bindings/wallet-wasm-js/js/index.d.ts]: fixed

File: src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs:518]: Agreed, will fix this

[the remark this answers is not in the document]
Comment (Mr-Leshiy) [reply src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs]: fixed

File: src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs [hunk 5]
```diff
@@ -513,14 +522,14 @@ impl FragmentReplayer {
             .get_mut(&address)
             .ok_or_else(|| ReplayError::NonVotingAccount(address.to_string()))?;
 
+        let secret_key = wallet.secret_key();
         // unwrap checked in the validation step
```
Comment (minikin) [inline src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs:526]: We don't need this `// unwrap checked in the validation step` comment anymore 🤓

File: src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs [hunk 6]
```diff
@@ -564,14 +573,19 @@ impl FragmentReplayer {
         };
 
         warn!("replaying a plain transaction from {} to {:?} with value {}, this is not coming from the app, might want to look into this", identifier, output_address, output.value);
+        let secret_key = wallet.secret_key();
         // unwrap checked in the validation step
         let builder_help = wallet
```
Comment (minikin) [inline src/catalyst-toolbox/catalyst-toolbox/src/recovery/tally.rs:578]: Can we remove the double unwrap() here too?

**Review by minikin — CHANGES_REQUESTED** [review]

**Review by minikin — APPROVED** [review]
```
██╗      ██████╗ ████████╗███╗   ███╗
██║     ██╔════╝ ╚══██╔══╝████╗ ████║
██║     ██║  ███╗   ██║   ██╔████╔██║
██║     ██║   ██║   ██║   ██║╚██╔╝██║
███████╗╚██████╔╝   ██║   ██║ ╚═╝ ██║
╚══════╝ ╚═════╝    ╚═╝   ╚═╝     ╚═╝
```

**Pull request merged**